Fsview extensions

mathstuf opened the pull request:
---
This is a rebase of #3227 that should be a lot easier to review. Basically, extension availability for views is now computed based on the view itself rather than looking globally.

Cc: @scheibelp

healther commented:
We are aware of an usability problem with the current state of #3227 due to the fact that some packages ship their own version of required libraries in the same position as "system" libraries.

One example is ```llvm``` which provides its own ```lib/python2.7/site-packages/six.py``` which conflicts with the ```lib/python2.7/site-packages/six.py``` from ```py-six```. When adding both ```llvm``` and ```py-six``` to the same view these files collide.
In the current state of #3227 this ends in an exception and the rest of the linking is aborted. As the "list "of packages to add is actually a set, the order of the linking is not guaranteed and we ended up adding an ignore-all-conflicts-by-skipping-existing-packages flag to the ```spack view add``` parser. Unfortunately we haven't managed to put that in #3227 yet.

We also thought about implementing some smarter version, i.e. deciding automatically whether to replace, skip or merge two files. However ```llvm/six.py``` and ```py-six/six.py``` where identical only up to formatting. So here it would have been save to choose either one, but we were uncomfortable with an automatic solution, as the content of the two files where not identical.
So we avoided the problem by just ignoring it and telling the people using it: Keep your eyes open, there might be inconsistent dependencies in this view. So far we haven't had any complaints...

cc @obreitwi

mathstuf commented:
LLVM should vendor it's six.py install to not conflict. That is a bug with LLVM that should be addressed.

scheibelp commented:
There appears to be enough support for ignoring conflicts that the view command would only need to add an `--ignore-conflicts` option; there appears to even be some code which looks for this option. When it's enabled it would print warnings when paths conflict (instead of failing).

healther commented:
If I recall correctly that only handled the "loading" of packages correctly but not the "activation" of packages

mathstuf commented:
I've added code which:

  - fixes `spack extensions` listing of installed packages: it was just listing activated packages;
  - the `extensions`, `activate`, and `deactivate` commands now take a `--view` (or `-v`) argument to operate on a view; and
  - `spack extensions --show <packages|installed|activated|all>` is now implemented.

healther commented on lib/spack/spack/database.py:
```diff
@@ -748,7 +748,23 @@ def installed_extensions_for(self, extendee_spec):
+                    yield spec.package
+            except spack.directory_layout.NoSuchExtensionError:
+                continue
+            # TODO: conditional way to do this instead of catching exceptions
```
why is catching exceptions a bad idea?

mathstuf replied:
This is copied from the old implementation. I imagine what is wanted is that instead of throwing an exception when a package doesn't extend a given spec is to return `False` or some other well-typed return value instead.

healther replied:
Ah sorry, I mistook the code position. I actually have no idea where this comes into play. However in general I prefer python code that indicates violated assumptions by exceptions instead of using a special invalid marker of the normal return value. I'm not too sure why the exception is caught in the first place (and especially swallowed without a logged event..)

mathstuf replied:
Looking at the code that does this, it really should return a bool. Currently it throws or it doesn't.

mathstuf replied:
And this code doesn't need the catch, it is `check_activated` that does the weird control flow.

mathstuf commented:
I found the bit which needed fixed for testing to work (a new variable added on this PR needed mocked out as well). Seems I need to fix some Python3 issues.

mathstuf commented:
Fixed the Python3 error. The YAML dumper needed to be told what encoding to use explicitly.

scheibelp commented on lib/spack/spack/filesystem_view.py:
```diff
@@ -0,0 +1,524 @@
+
+        set(map(self._check_no_ext_conflicts, extensions))
+        # fail on first error, otherwise link extensions as well
+        if all(map(self.add_standalone, standalones)):
```
I'm starting to look at this now. I noticed if one of the packages is external and has a prefix like:

```
packages:
    openssl:
        paths:
            openssl@1.0.1e: /usr/local
```

That all of `/usr/local` will be added. Could you add a check to skip external specs here?

mathstuf replied:
Ew, yeah, I'll do that.

mathstuf replied:
OK, fixed and I've added a test for it.

scheibelp commented on lib/spack/spack/cmd/view.py:
```diff
@@ -48,258 +48,187 @@
+        # status and remove can map the name to packages in view
+        specs = relaxed_disambiguate(args.specs, view)
+
+    activated = list(filter(lambda s: s.package.is_extension and
```
The main goal of collecting these is to avoid a case where an extendee package has globally activated extensions and we add it to the view. I think a stronger check here would be to grab the extendee package and check `activated_extensions_for`. This would also mean changing the checks in `test_view_extension_global_activation`.

scheibelp commented on lib/spack/spack/test/cmd/view.py:
```diff
@@ -0,0 +1,91 @@
+    install('extension1@2.0')
+    install('extension2@1.0')
+    viewpath = str(tmpdir.mkdir('view'))
+    view('symlink', viewpath, 'extension1@1.0')
```
IMO it is also worth confirming that the files in the `extension1` package were created in the view prefix.

scheibelp commented on lib/spack/spack/test/cmd/view.py:
```diff
@@ -0,0 +1,91 @@
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack.main import SpackCommand
```
Could you add a test for removing extensions from views?

IMO it would also be useful to add a test for a file conflict (two different packages have the same file and are added to a view).

mathstuf commented:
I fixed the symlinking an extendee with global activations and added the other tests as well.

scheibelp commented:
All changes appear good. However, I was looking through packages which implement `activate` and realized that `perl` still refers to `spack.store.layout.extension_map`.  I'm not sure why it does because `Package.activate` already handles modifying the extension map, so I think the references can just be removed.

mathstuf commented:
Python does the same thing; I just coverted it over to using the same argument setup.

scheibelp commented:
The `aspell` package is also extendable and implements `activate` in a way that only works for globals. 

This may not matter though if `aspell` doesn't have a `PYTHONPATH` equivalent. @hartzell (as the contributor of Spack's `aspell` package definition) do you happen to know if it does.

mathstuf commented:
Not sure about `aspell`, but `hunspell` supports `DICPATH`. Looking at `aspell` source code, I don't see one.

mathstuf commented:
Yeah, the best I see with `aspell` is a configuration file pointed at with `ASPELL_CONF`.

scheibelp commented:
> the best I see with aspell is a configuration file pointed at with ASPELL_CONF.

OK - I think in that case it would be worthwhile if `Aspell.activate` failed for a `YamlViewExtensionsLayout`, or if `Aspell.activate` used the destination directory given by `aspell dump config dict-dir` but replaced the spack prefix with the view prefix (to be used by `ASPELL_CONF`). At the moment if I do

```spack view soft view1 aspell6-en```

Then it adds `aspell6-en` to the spack prefix silently.

mathstuf commented:
`aspell` now rejects working with non-global extensions.

scheibelp commented:
@mathstuf the following patch adds tests for python activation (since `Python.activate` implementation is distinct from `Package.activate`): https://gist.github.com/scheibelp/f106ce5fcfcb6a5a55ea32dc9dc3018a

Does that seem reasonable? In the case of `perl` a test like `test_python_activation` would have caught https://github.com/LLNL/spack/pull/5415#issuecomment-339858670.

mathstuf commented:
Yeah, the test looks sensible. I'll add cases for Perl as well.

scheibelp commented on var/spack/repos/builtin/packages/aspell/package.py:
```diff
@@ -47,6 +47,12 @@ class Aspell(AutotoolsPackage):
+        extensions_layout = args.get("extensions_layout",
+                                     spack.store.extensions)
+        if extensions_layout is not spack.store.extensions:
+            raise ExtensionError(
```
Also need to import ExtensionError

scheibelp commented on var/spack/repos/builtin/packages/aspell/package.py:
```diff
@@ -47,6 +47,12 @@ class Aspell(AutotoolsPackage):
     #   - extension.prefix.lib instead of extension.prefix in LinkTree()
     #   - dest_dir instead of self.prefix in tree.(find_conflict|merge)()
     def activate(self, extension, **kwargs):
+        extensions_layout = args.get("extensions_layout",
```
`args` should be `kwargs`

scheibelp commented on var/spack/repos/builtin/packages/aspell/package.py:
```diff
@@ -62,6 +68,12 @@ def ignore(filename):
         tree.merge(dest_dir, ignore=ignore)
 
     def deactivate(self, extension, **kwargs):
+        extensions_layout = args.get("extensions_layout",
```
`args` should be `kwargs`

scheibelp commented:
These test failures do not appear related to the edits so I am closing and reopening to restart tests.

scheibelp commented:
@obreitwi thanks for your work on this!

@mathstuf thanks for the added work (including testing)!

@tgamblin FYI this is now merged

tgamblin commented:
Thanks everyone!

obreitwi commented:
@scheibelp @mathstuf Thanks for getting this done!

haampie commented on lib/spack/spack/cmd/extensions.py:
```diff
@@ -76,37 +102,46 @@ def extensions(parser, args):
+                     for s in spack.store.db.installed_extensions_for(spec)]
+
+        if show_all:
+            print
```
what was the idea here @mathstuf?

mathstuf replied:
Uhh…probably either stray debugging (unlikely) or an incomplete/untested flag combination.
